### PR TITLE
Шуравина Оксана. Задача 1. Вариант 21. Интегрирование – метод Монте-Карло.

### PR DESCRIPTION
**Pull request by shyravina19102004**

### CODE REVIEW

File: tasks/mpi/shuravina_o_monte_carlo/func_tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  ASSERT_EQ(testMpiTaskParallel->validation(), true);
+  testMpiTaskParallel->pre_processing();
+  testMpiTaskParallel->run();
+  testMpiTaskParallel->post_processing();
+
+  if (world.rank() == 0) {
+    double expected_integral = 1.0 / 3.0;
+    ASSERT_NEAR(expected_integral, out[0], 0.01);
+  }
+}
```
Comment (dasha1112) [inline tasks/mpi/shuravina_o_monte_carlo/func_tests/main.cpp:31]: Please, add more then one test

File: tasks/seq/shuravina_o_monte_carlo/func_tests/main.cpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  auto testTaskSequential = std::make_shared<shuravina_o_monte_carlo::MonteCarloIntegrationTaskSequential>(taskDataSeq);
+  ASSERT_EQ(testTaskSequential->validation(), true);
+  testTaskSequential->pre_processing();
+  testTaskSequential->run();
+  testTaskSequential->post_processing();
+
+  double expected_integral = 1.0 / 3.0;
+  ASSERT_NEAR(expected_integral, out[0], 0.01);
+}
```
Comment (dasha1112) [inline tasks/seq/shuravina_o_monte_carlo/func_tests/main.cpp:23]: Please, add more then one test there too

Comment (shyravina19102004) [reply tasks/seq/shuravina_o_monte_carlo/func_tests/main.cpp:41]: пофиксила

File: tasks/mpi/shuravina_o_monte_carlo/func_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+  testMpiTaskParallel->run();
+  testMpiTaskParallel->post_processing();
+
+  if (world.rank() == 0) {
+    double expected_integral = 1.0 / 3.0;
+    ASSERT_NEAR(expected_integral, out[0], 0.01);
+  }
+
+  int num_processes = world.size();
+  int rank = world.rank();
```
Comment (keshaproger) [inline tasks/mpi/shuravina_o_monte_carlo/func_tests/main.cpp:81]: Unused variable, remove it

**Review by keshaproger — CHANGES_REQUESTED** [review]

File: tasks/seq/shuravina_o_monte_carlo/include/ops_seq.hpp [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  bool validation() override;
+  bool run() override;
+  bool post_processing() override;
+
+  void set_interval(double a, double b) {
+    a_ = a;
+    b_ = b;
+  }
+
+  void set_function(std::function<double(double)> func) { f_ = func; }
```
Comment (keshaproger) [inline tasks/seq/shuravina_o_monte_carlo/include/ops_seq.hpp:26]: Change: void set_function(std::function<double(double)> func) { f_ = func; }
to: void set_function(std::function<double(double)> func) { f_ = std::move(func); }

**Review by keshaproger — CHANGES_REQUESTED** [review]

File: tasks/seq/shuravina_o_monte_carlo/include/ops_seq.hpp (hunk 4 above)
Comment (shyravina19102004) [reply tasks/seq/shuravina_o_monte_carlo/include/ops_seq.hpp]: fixed

**Review by keshaproger — APPROVED** [review]

File: tasks/mpi/shuravina_o_monte_carlo/src/ops_mpi.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  internal_order_test();
+  int num_processes = world.size();
+  int rank = world.rank();
+
+  auto f = [](double x) { return x * x; };
+
+  double a = 0.0;
+  double b = 1.0;
+
+  int num_points = 1000000;
```
Comment (aobolensk) [inline tasks/mpi/shuravina_o_monte_carlo/src/ops_mpi.cpp:38]: This should be customizable

File: tasks/mpi/shuravina_o_monte_carlo/src/ops_mpi.cpp [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  }
+  return true;
+}
+
+bool shuravina_o_monte_carlo::MonteCarloIntegrationTaskParallel::run() {
+  internal_order_test();
+  int num_processes = world.size();
+  int rank = world.rank();
+
+  auto f = [](double x) { return x * x; };
```
Comment (aobolensk) [inline tasks/mpi/shuravina_o_monte_carlo/src/ops_mpi.cpp:33]: Is function hardcoded here? This is not correct. You need to support any integrable function here

File: tasks/mpi/shuravina_o_monte_carlo/src/ops_mpi.cpp [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+  double a = 0.0;
+  double b = 1.0;
```
Comment (aobolensk) [inline tasks/mpi/shuravina_o_monte_carlo/src/ops_mpi.cpp:36]: This also should be cuatomizable

File: tasks/mpi/shuravina_o_monte_carlo/perf_tests/main.cpp [hunk 8]
```diff
@@ -0,0 +1,69 @@
+#include <gtest/gtest.h>
+
+#include <boost/mpi/timer.hpp>
+#include <vector>
+
+#include "core/perf/include/perf.hpp"
+#include "mpi/shuravina_o_monte_carlo/include/ops_mpi.hpp"
+
+TEST(MonteCarloIntegrationTaskParallel, test_pipeline_run) {
```
Comment (aobolensk) [inline tasks/mpi/shuravina_o_monte_carlo/perf_tests/main.cpp:9]: Perf test should be run on more iterations (they are called num_points in your code)

**Review by aobolensk — CHANGES_REQUESTED** [review]

**Comment by aobolensk** [comment]
Previous approves are not considered valid. Please collect 1 more peer review on top

File: tasks/mpi/shuravina_o_monte_carlo/perf_tests/main.cpp (hunk 8 above)
Comment (shyravina19102004) [reply tasks/mpi/shuravina_o_monte_carlo/perf_tests/main.cpp:9]: fixed

File: tasks/mpi/shuravina_o_monte_carlo/src/ops_mpi.cpp [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+bool shuravina_o_monte_carlo::MonteCarloIntegrationTaskParallel::run() {
+  internal_order_test();
+  int num_processes = world.size();
+  int rank = world.rank();
+
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::uniform_real_distribution<> dis(a_, b_);
```
Comment (allnes) [inline tasks/mpi/shuravina_o_monte_carlo/src/ops_mpi.cpp:35]: please initilize it outside run()

Comment (shyravina19102004) [reply tasks/mpi/shuravina_o_monte_carlo/src/ops_mpi.cpp]: fixed

**Review by dasha1112 — APPROVED** [review]

File: tasks/mpi/shuravina_o_monte_carlo/src/ops_mpi.cpp (hunk 6 above)
Comment (aobolensk) [reply tasks/mpi/shuravina_o_monte_carlo/src/ops_mpi.cpp]: Tests still contain `x * x` only

File: tasks/mpi/shuravina_o_monte_carlo/func_tests/main.cpp [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    int local_num_points = num_points / num_processes;
+
+    std::vector<int> local_points_count(num_processes, 0);
+    boost::mpi::all_gather(world, local_num_points, local_points_count);
+
+    for (int i = 0; i < num_processes; ++i) {
+      ASSERT_EQ(local_points_count[i], local_num_points);
+    }
+  } catch (const std::exception& e) {
+    std::cerr << "Process " << world.rank() << " caught exception: " << e.what() << std::endl;
```
Comment (aobolensk) [inline tasks/mpi/shuravina_o_monte_carlo/func_tests/main.cpp:121]: Please, do not print in tests

File: tasks/mpi/shuravina_o_monte_carlo/func_tests/main.cpp [hunk 11]
```diff
@@ -0,0 +1,219 @@
+#include <gtest/gtest.h>
+
+#include <boost/mpi/collectives.hpp>
+#include <boost/mpi/timer.hpp>
+#include <vector>
+
+#include "core/perf/include/perf.hpp"
+#include "mpi/shuravina_o_monte_carlo/include/ops_mpi.hpp"
+TEST(MonteCarloIntegrationTaskParallel, Test_Integration) {
```
Comment (aobolensk) [inline tasks/mpi/shuravina_o_monte_carlo/func_tests/main.cpp:9]: Too few tests, you have only one function tested

**Review by aobolensk — CHANGES_REQUESTED** [review]

**Comment by aobolensk** [comment]
> ## [Codecov](https://app.codecov.io/gh/learning-process/ppc-2024-autumn/pull/266?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=learning-process) Report
> Attention: Patch coverage is `80.59701%` with `13 lines` in your changes missing coverage. Please review.
> 
> > Project coverage is 96.34%. Comparing base [(`5daf9ab`)](https://app.codecov.io/gh/learning-process/ppc-2024-autumn/commit/5daf9ab13b05a0ed91c1162cc013d1e8571de199?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=learning-process) to head [(`e7b2ef2`)](https://app.codecov.io/gh/learning-process/ppc-2024-autumn/commit/e7b2ef2e0383775c72e2b8b3c28ddc8ea6cca1c6?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=learning-process).
> > Report is 168 commits behind head on master.
> 
> [Files with missing lines](https://app.codecov.io/gh/learning-process/ppc-2024-autumn/pull/266?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=learning-process)	Patch %	Lines
> [tasks/mpi/shuravina_o_monte_carlo/src/ops_mpi.cpp](https://app.codecov.io/gh/learning-process/ppc-2024-autumn/pull/266?src=pr&el=tree&filepath=tasks%2Fmpi%2Fshuravina_o_monte_carlo%2Fsrc%2Fops_mpi.cpp&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=learning-process#diff-dGFza3MvbXBpL3NodXJhdmluYV9vX21vbnRlX2NhcmxvL3NyYy9vcHNfbXBpLmNwcA==)	78.57%	[3 Missing and 3 partials ⚠️ ](https://app.codecov.io/gh/learning-process/ppc-2024-autumn/pull/266?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=learning-process)
> [tasks/seq/shuravina_o_monte_carlo/src/ops_seq.cpp](https://app.codecov.io/gh/learning-process/ppc-2024-autumn/pull/266?src=pr&el=tree&filepath=tasks%2Fseq%2Fshuravina_o_monte_carlo%2Fsrc%2Fops_seq.cpp&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=learning-process#diff-dGFza3Mvc2VxL3NodXJhdmluYV9vX21vbnRlX2NhcmxvL3NyYy9vcHNfc2VxLmNwcA==)	79.16%	[2 Missing and 3 partials ⚠️ ](https://app.codecov.io/gh/learning-process/ppc-2024-autumn/pull/266?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=learning-process)
> [...ks/mpi/shuravina_o_monte_carlo/include/ops_mpi.hpp](https://app.codecov.io/gh/learning-process/ppc-2024-autumn/pull/266?src=pr&el=tree&filepath=tasks%2Fmpi%2Fshuravina_o_monte_carlo%2Finclude%2Fops_mpi.hpp&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=learning-process#diff-dGFza3MvbXBpL3NodXJhdmluYV9vX21vbnRlX2NhcmxvL2luY2x1ZGUvb3BzX21waS5ocHA=)	87.50%	[1 Missing ⚠️ ](https://app.codecov.io/gh/learning-process/ppc-2024-autumn/pull/266?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=learning-process)
> [...ks/seq/shuravina_o_monte_carlo/include/ops_seq.hpp](https://app.codecov.io/gh/learning-process/ppc-2024-autumn/pull/266?src=pr&el=tree&filepath=tasks%2Fseq%2Fshuravina_o_monte_carlo%2Finclude%2Fops_seq.hpp&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=learning-process#diff-dGFza3Mvc2VxL3NodXJhdmluYV9vX21vbnRlX2NhcmxvL2luY2x1ZGUvb3BzX3NlcS5ocHA=)	85.71%	[1 Missing ⚠️ ](https://app.codecov.io/gh/learning-process/ppc-2024-autumn/pull/266?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=learning-process)
> Additional details and impacted files
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/learning-process/ppc-2024-autumn/pull/266?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=learning-process). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=learning-process).
> 
> 🚨 Try these New Features:
> 
> * [Flaky Tests Detection](https://docs.codecov.com/docs/test-result-ingestion-beta) - Detect and resolve failed and flaky tests

It seems your code coverage is too low

File: tasks/mpi/shuravina_o_monte_carlo/func_tests/main.cpp (hunk 10 above)
Comment (shyravina19102004) [reply tasks/mpi/shuravina_o_monte_carlo/func_tests/main.cpp]: fixed

**Pull request closed without merging**